core/api/container/src/main/java/org/codehaus/cargo/container/spi/deployer/AbstractJsr88Deployer.java

```diff
@@ -423,35 +426,48 @@ else if (DeployableType.WAR.equals(deployable.getType()))
                 + " has no JSR-88 match and cannot be remotely deployed.");
         }
 
-        TargetModuleID targetModule = null;
         TargetModuleID[] modules = deploymentManager.getAvailableModules(moduleType,
             deploymentManager.getTargets());
 
         StringBuilder sb = new StringBuilder();
         sb.append("Cannot find the module \"");
         sb.append(moduleName);
         sb.append("\". Available modules:");
-        for (TargetModuleID module : modules)
-        {
+        
+        List<TargetModuleID> targetModules = new ArrayList<TargetModuleID>();    
+
+        for (TargetModuleID module : modules) {
             String moduleId = module.getModuleID();
 
-            if (moduleName.equals(moduleId))
-            {
-                targetModule = module;
-                break;
+            if (moduleName.equals(moduleId)) {
+                targetModules.add(module);
+            } else {
+                sb.append("\n\t- ");
+                sb.append(moduleId);
             }
-            sb.append("\n\t- ");
-            sb.append(moduleId);
         }
 
-        if (targetModule == null)
-        {
+        if (targetModules.isEmpty()) {
             throw new CargoException(sb.toString());
         }
-
-        TargetModuleID[] targetModules = new TargetModuleID[1];
-        targetModules[0] = targetModule;
-        return targetModules;
+        
+        List<String> currentTargets = Arrays.asList(this.configuration.getProperties().get("cargo.glassfish.target").trim().split(","));
+        List<TargetModuleID> finalTargetModules = new ArrayList<TargetModuleID>();
+
+        if (currentTargets.isEmpty()) {
+            finalTargetModules = targetModules;
+        } else {
+            for (TargetModuleID targetModule : targetModules) {
+                for (String target : currentTargets) {
+                    if (targetModule.getTarget().getName().equals(target)) {
+                        finalTargetModules.add(targetModule);
+                        break;
+                    }
+                }
+            }
+        }
+     
+        return finalTargetModules.toArray(new TargetModuleID[finalTargetModules.size()]);
```

Comment:
as of JDK7, `finalTargetModules.toArray(new TargetModuleID[**0**]);` is preferred.